Added snake case naming policy to the JSON serializer

YohDeadfall opened the pull request:
Fixes #39564. The code is written some time ago for our ADO.NET driver for PostgreSQL where snake case is the common and default naming style.

YohDeadfall committed:
Added snake case naming policy

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,53 @@
+    public static class SnakeCaseUnitTests
+    {
+        [Fact]
+        public static void ToSnakeCaseTest()
```
Based on discussion from the API review, we should add tests for non-ascii text and for text with surrogate pairs (including invalid surrogate pairs) and also validate (at least locally - but I would be ok with checking it in) that the output matches Newtonsoft.Json.

I was working on some testing for CamelCase, so you could try and leverage that as a starting point. Note that for non-ascii characters, we wouldn't match Newtonsoft.Json exactly due to the default escaping policy in S.T.Json (feel free to either override that or loosen the assert condition for the tests).
```C#
[Theory]
[InlineData("lowerCase")]
[InlineData("UpperCase")]
[InlineData("UPperCase")]
[InlineData("lower Case")]
[InlineData("loweR Case")]
[InlineData("Upper Case")]
[InlineData("UppeR Case")]
[InlineData("lower case")]
[InlineData("loweR case")]
[InlineData("Upper case")]
[InlineData("UppeR case")]
[InlineData("lower cASe")]
[InlineData("loweR cASe")]
[InlineData("Upper cASe")]
[InlineData("UppeR cASe")]
[InlineData("UPper Case")]
[InlineData("UPpeR Case")]
[InlineData("UPper case")]
[InlineData("UPpeR case")]
[InlineData("UPper cASe")]
[InlineData("UPpeR cASe")]

[InlineData("i")]
[InlineData("I")]
[InlineData("i ")]
[InlineData("I ")]

[InlineData("ii")]
[InlineData("II")]
[InlineData("iI")]
[InlineData("Ii")]
[InlineData("ii ")]
[InlineData("II ")]
[InlineData("iI ")]
[InlineData("Ii ")]

[InlineData("iii")]
[InlineData("III")]
[InlineData("iiI")]
[InlineData("iIi")]
[InlineData("Iii")]
[InlineData("iII")]
[InlineData("IIi")]
[InlineData("IiI")]
[InlineData("iii ")]
[InlineData("III ")]
[InlineData("iiI ")]
[InlineData("iIi ")]
[InlineData("Iii ")]
[InlineData("iII ")]
[InlineData("IIi ")]
[InlineData("IiI ")]

[InlineData("iPhone")]
[InlineData("IPhone")]
[InlineData("IPHone")]
[InlineData("IPH one")]
[InlineData("IPH One")]
[InlineData("IPHONe")]
[InlineData("IPHON e")]
[InlineData("IPHON E")]
[InlineData("IPHONE")]
public static void TestCasing(string baseString)
{
    var testObject = new TestClassWithDictionary
    {
        Dictionary = new Dictionary<string, string>
        {
            { baseString, "hi" }
        }
    };

    string json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
    // Assert JSON output is correct

    testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
    json = JsonSerializer.Serialize(testObject);
    // Assert JSON output is correct after round-tripping

    DefaultContractResolver contractResolver = new DefaultContractResolver
    {
        NamingStrategy = new CamelCaseNamingStrategy()
    };

    string expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
    // Assert and compare against Newtonsoft.Json

    char[] baseStringArray = baseString.ToCharArray();
    char[] testCharArray = new char[baseStringArray.Length + 2];
    for (int i = 0; i <= baseStringArray.Length; i++)
    {
        for (int j = 0; j < i; j++)
        {
            testCharArray[j] = baseStringArray[j];
        }
        testCharArray[i] = '\uD835';
        testCharArray[i + 1] = '\uDD80';
        //testCharArray[i] = '\uD801';
        //testCharArray[i + 1] = '\uDC27';
        for (int j = i; j < baseStringArray.Length; j++)
        {
            testCharArray[j + 2] = baseStringArray[j];
        }

        string testString = new string(testCharArray);

        testObject = new TestClassWithDictionary
        {
            Dictionary = new Dictionary<string, string>
            {
                { testString, "hi" }
            }
        };

        json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
        // Assert JSON output is correct

        testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
        json = JsonSerializer.Serialize(testObject);
        // Assert JSON output is correct after round-tripping

        expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
        // Assert and compare against Newtonsoft.Json
    }

    baseStringArray = baseString.ToCharArray();
    testCharArray = new char[baseStringArray.Length + 1];
    for (int i = 0; i <= baseStringArray.Length; i++)
    {
        for (int j = 0; j < i; j++)
        {
            testCharArray[j] = baseStringArray[j];
        }
        testCharArray[i] = '\uD835';
        for (int j = i; j < baseStringArray.Length; j++)
        {
            testCharArray[j + 1] = baseStringArray[j];
        }

        string testString = new string(testCharArray);

        testObject = new TestClassWithDictionary
        {
            Dictionary = new Dictionary<string, string>
            {
                { testString, "hi" }
            }
        };

        json = JsonSerializer.Serialize<TestClassWithDictionary>(testObject, new JsonSerializerOptions { DictionaryKeyPolicy = JsonNamingPolicy.CamelCase });
        // Assert JSON output is correct

        testObject = JsonSerializer.Deserialize<TestClassWithDictionary>(json);
        json = JsonSerializer.Serialize(testObject);
        // Assert JSON output is correct after round-tripping

        expected = JsonConvert.SerializeObject(testObject, new JsonSerializerSettings { ContractResolver = contractResolver });
        // Assert and compare against Newtonsoft.Json
    }
}
```

YohDeadfall replied:
Newtonsoft.Json had some bad translations (`Xml2Json` to `xml2_json`) and the bug I mentioned. Should the behavior be the same or should it be improved a bit? Maybe open an issue in Newtonsoft.Json?

/cc @JamesNK

ahsonkhan replied:
> Newtonsoft.Json had some bad translations (Xml2Json to xml2_json) and the bug I mentioned.

It's fine for those cases for us to deviate, but for other test cases making sure our output matches would be good (or better yet, codify the expected output from Newtonsoft.Json into test cases and verify our output matches against those strings).

> Should the behavior be the same or should it be improved a bit?

In cases where the Newtonsoft behavior is buggy/undesirable, we should certainly try to make it better here (possibly in both libraries). 

> Maybe open an issue in Newtonsoft.Json?

I'd leave the Newtonsoft.Json side of things to @JamesNK, but filing an issue wouldn't hurt :)

YohDeadfall replied:
Already tracked by JamesNK/Newtonsoft.Json#1956.

ahsonkhan replied:
> Based on discussion from the API review, we should add tests for non-ascii text and for text with surrogate pairs (including invalid surrogate pairs) and also validate (at least locally - but I would be ok with checking it in) that the output matches Newtonsoft.Json.

Can you add some more tests for non-ascii text?

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonNamingPolicy.cs:
```diff
@@ -19,6 +19,11 @@ public abstract class JsonNamingPolicy
         public static JsonNamingPolicy CamelCase { get; } = new JsonCamelCaseNamingPolicy();
 
+        /// <summary>
+        /// Returns the naming policy for snake-casing.
```
nit: Follow the content from the docs:
https://github.com/dotnet/dotnet-api-docs/blob/4f9dbb8474e1972f342b252a8d207e1155f41479/xml/System.Text.Json/JsonNamingPolicy.xml#L54-L55
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonnamingpolicy.camelcase?view=netcore-3.0

So:
```suggestion
        /// Gets the naming policy for snake-casing.
```

And add a value xml tag as well (which is equivalent to a "return" tag but for property getters).

YohDeadfall replied:
Should I do the same for camel case?

ahsonkhan replied:
> Should I do the same for camel case?

Sure, if you like. It wouldn't hurt doing a minor cleanup while you are in the similar area for this case. However, we currently don't back-port updates from the docs to the source since docs is the source of truth and we'd continuously get out-of-sync.

YohDeadfall replied:
Do you mean that documentation is written here first, than moved to the docs repo and never updated in the code? Just trying to understand the flow.

ahsonkhan replied:
> Do you mean that documentation is written here first, than moved to the docs repo and never updated in the code? Just trying to understand the flow.

Yes. The docs repo is the source of truth and contains "live" changes, along with review done by the doc writers during the initial move for correctness in terms of format/consistency/following guidelines. We don't currently have a great way to bring those changes back to code.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,72 @@
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
```
Any particular reason to use a different implementation than what was done in Newtonsoft.Json (which this PR had: https://github.com/dotnet/corefx/pull/40003)?

YohDeadfall replied:
Npgsql used the same code as Newtonsoft.Json, but it had a bug which I fixed. See npgsql/npgsql#2152.

ahsonkhan commented on src/System.Text.Json/tests/System.Text.Json.Tests.csproj:
```diff
@@ -69,6 +69,7 @@
     <Compile Include="Serialization\PropertyNameTests.cs" />
     <Compile Include="Serialization\PropertyVisibilityTests.cs" />
     <Compile Include="Serialization\ReadValueTests.cs" />
+    <Compile Include="Serialization\SnakeCaseUnitTests.cs" />
```
It will be good to bring in some of the tests from this PR as well:
https://github.com/dotnet/corefx/pull/40003

stephentoub commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,72 @@
+                return name;
+            }
+
+            const char underscore = '_';
```
Does this really need to be in a named const?  Using '_' is shorter, and at least to me just as clear if not more so.

If we must stick with the named const, it should be Underscore instead of underscore.

YohDeadfall replied:
Should it be named as `Separator` instead or inline the value instead?

stephentoub commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,72 @@
+            }
+
+            const char underscore = '_';
+            const UnicodeCategory noneCategory = UnicodeCategory.Control;
```
What does the "none" here mean?

Should also be "NoneCategory" rather than "noneCategory".

YohDeadfall replied:
It's a trick to make `previousCategory` non-nullable which is based on that any character from not mentioned in the method Unicode categories are replaced an underscore. Therefore, letters, decimals and control characters are preserved.

Maybe it should named better...

stephentoub commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,72 @@
+            const char underscore = '_';
+            const UnicodeCategory noneCategory = UnicodeCategory.Control;
+
+            StringBuilder builder = new StringBuilder();
```
This is going to add a fair amount of allocation per conversion.

YohDeadfall replied:
Oops, a good catch. Is it okay to use the following capacity?

```csharp
var capacity = name.Length + Math.Min(2, name.Length / 5);
// 5 is a median word length in English
```

VassilAtanasov commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonNamingPolicy.cs:
```diff
@@ -19,6 +19,11 @@ public abstract class JsonNamingPolicy
+        /// </summary>
+        public static JsonNamingPolicy SnakeCase { get; } = new JsonSnakeCaseNamingPolicy();
+
         internal static JsonNamingPolicy Default { get; } = new JsonDefaultNamingPolicy();
```
Why this is internal? Developers need to be able to use it pass through PascalCase

YohDeadfall replied:
A good question, but unrelated to this PR. Makes sense to open an issue so anyone would be able to find the answer.

VassilAtanasov commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -227,6 +227,7 @@ public abstract partial class JsonNamingPolicy
         protected JsonNamingPolicy() { }
         public static System.Text.Json.JsonNamingPolicy CamelCase { get { throw null; } }
+        public static System.Text.Json.JsonNamingPolicy SnakeCase { get { throw null; } }
         public abstract string ConvertName(string name);
```
I agree it is not related to the PR but I don't know how to propose this fix properly ...
Also a need:     
```
/// <summary>
/// Property names are not changed
/// </summary>
public static System.Text.Json.JsonNamingPolicy Default{ get { throw null; } }
```

ahsonkhan commented:
> For me it looks like the current behavior is better, so makes sense to remove failing tests. Are you okay with this, @ahsonkhan?

I agree. @JamesNK - what do you think about those cases? I think having a single `_` per space character is more appropriate than having multiple (and similar concerns when the last character is a space). I don't know if we should remain 100% compatible with the underscore casing output (even for these cases), and I'd like to hear what compelled you to have the showcased behavior in Json.NET.

YohDeadfall commented:
/azp run

YohDeadfall commented:
@ahsonkhan There are infrastructure related failures again...

YohDeadfall commented:
/azp run

YohDeadfall commented:
@ahsonkhan All checks passed.

lpereira commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,67 @@
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            StringBuilder builder = new StringBuilder(name.Length + Math.Min(2, name.Length / 5));
```
Would this benefit from `ValueStringBuilder` with a `stackalloc` buffer (alocated with a size cap)?

YohDeadfall replied:
Sure, but in that case `ValueStringBuilder` should be referenced by the project since the struct isn't public yet (see #28379). If it's okay, I will do it (:

/cc @stephentoub @ahsonkhan

ahsonkhan replied:
Since `ValueStringBuilder` is internal (within corelib), I wouldn't take a dependency on it here. I don't think it's worth it in this context, particularly because creating the converted name only happens once and then the result is stored.

Though I believe it is feasible to do, if we really wanted to. For example:
https://github.com/dotnet/corefx/blob/c78b7189eabe20c6f4c8d386ca18b68fea513189/src/System.Runtime.Extensions/src/System/Net/WebUtility.cs#L58

The main benefit would be on startup, but we might be able to write this method without using `ValueStringBuilder` and still be faster. I am heads down on some other work currently, and will review this change soon. Thanks for your patience, @YohDeadfall, appreciate it.

stephentoub replied:
> I don't think it's worth it in this context, particularly because creating the converted name only happens once and then the result is stored

@ahsonkhan, are there circumstances where it might happen for every occurrence? e.g. when happens when serializing a dictionary?

ahsonkhan replied:
I will take a second look and verify, but from what I understand, calls to `ConvertName` should only ever happen once. @steveharter, @layomia - feel free to chime in about the behavior of serializing dictionary keys/values.

stephentoub replied:
> from what I understand, calls to ConvertName should only ever happen once

That doesn't seem to be the case, e.g.
https://github.com/dotnet/corefx/blob/1338e4e8a542c3b3e20d99e544e09860f23e99ea/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs#L154
but maybe I'm just missing where that gets cached.  I wouldn't expect it to be cached, though, considering arbitrary dictionaries might get serialized, and their keys could be all over the map.

ahsonkhan replied:
Good point @stephentoub. I was looking at `PropertyNamingPolicy` only, but `JsonNamingPolicy` applies to `DictionaryKeyPolicy` as well.

Once the implementation of `JsonSnakeCaseNamingPolicy` is well tested, we should consider trying to optimize the built-in naming policies. This can be done in a follow-up PR (no point trying to optimize it without a larger test set that ensures correctness). Filed an issue: https://github.com/dotnet/corefx/issues/42581

``` ini

BenchmarkDotNet=v0.11.5.1159-nightly, OS=Windows 10.0.18362
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.100-preview1-014459
  [Host]     : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT
  Job-HPZZRZ : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250.0000 ms  MaxIterationCount=10  
MinIterationCount=5  WarmupCount=1  

```
|  Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
| Default | 2.426 us | 0.0473 us | 0.0281 us | 2.425 us | 2.391 us | 2.473 us |  1.00 |    0.00 | 0.1567 |     - |     - |     656 B |
|   Camel | 3.040 us | 0.0895 us | 0.0592 us | 3.027 us | 2.979 us | 3.150 us |  1.26 |    0.04 | 0.2491 |     - |     - |    1048 B |
|   Snake | 4.293 us | 0.0939 us | 0.0621 us | 4.280 us | 4.195 us | 4.397 us |  1.77 |    0.01 | 0.5624 |     - |     - |    2432 B |


```C#
public class SerializeDictionaryWithNamingPolicy
{
    JsonSerializerOptions _camelOptions;
    JsonSerializerOptions _snakeOptions;

    Dictionary<string, string> _foo;

    [GlobalSetup]
    public void Setup()
    {
        _camelOptions = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase, DictionaryKeyPolicy = JsonNamingPolicy.CamelCase };
        _snakeOptions = new JsonSerializerOptions { PropertyNamingPolicy = new JsonSnakeCaseNamingPolicy(), DictionaryKeyPolicy = new JsonSnakeCaseNamingPolicy() };

        _foo = new Dictionary<string, string>()
        {
            ["ID"] = "A",
            ["url"] = "A",
            ["URL"] = "A",
            ["THIS IS SPARTA"] = "A",
            ["IsCIA"] = "A",
            ["iPhone"] = "A",
            ["IPhone"] = "A",
            ["xml2json"] = "A",
            ["already_snake_case"] = "A",
            ["IsJSONProperty"] = "A",
            ["ABCDEFGHIJKLMNOP"] = "A",
            ["Hi!! This is text.Time to test."] = "A",
        };
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark(Baseline = true)]
    public string Default()
    {
        return JsonSerializer.Serialize(_foo);
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark]
    public string Camel()
    {
        return JsonSerializer.Serialize(_foo, _camelOptions);
    }

    [BenchmarkCategory(Categories.CoreFX, Categories.JSON)]
    [Benchmark]
    public string Snake()
    {
        return JsonSerializer.Serialize(_foo, _snakeOptions);
    }
}
```

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

stephentoub commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,141 @@
+            Assert.Equal("building_property", ConvertToSnakeCase("BUILDING Property"));
+            Assert.Equal("building_property", ConvertToSnakeCase("Building Property"));
+            Assert.Equal("building_property", ConvertToSnakeCase("BUILDING PROPERTY"));
+        }
```
Nit: can this instead be:
```C#
[Theory]
[InputData("url_value", "URLValue")]
[InputData(...)] // the cases you have
public static void Convert_SpecifiedName_MatchesExpectedName(string expected, string input)
{
    Assert.Equal(expected, JsonNamingPolicy.SnakeCase.ConvertName(name));
}
```

YohDeadfall replied:
Yep, but I assumed that @ahsonkhan would like to implement it this way to compare behaviors. Are you okay with changing this, @ahsonkhan?

ahsonkhan replied:
Yep, the change seems reasonable to me. Tests showing the expected/actual string before/after is clear in both the current way the test is written and what @stephentoub suggested, so I am supportive of the change.

stephentoub commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,141 @@
+            JsonNamingPolicy policy = JsonNamingPolicy.SnakeCase;
+            string value = policy.ConvertName(name);
+            return value;
+        }
```
And then delete this helper.

stephentoub commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,67 @@
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            StringBuilder builder = new StringBuilder(name.Length + Math.Min(2, name.Length / 5));
```
Where did the `Math.Min(2, name.Length / 5)` formula come from?  A comment would be helpful.

YohDeadfall replied:
Oh, there is a typo here. There should be `Max` instead of `Min`. Fixed and comment added.

stephentoub reviewed (approved):
Other than my comments, LGTM.

danmoseley commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,141 @@
+        [Fact]
+        public static void Convert_SpecifiedName_MatchesExpectedName()
+        {
+            Assert.Equal("url_value", ConvertToSnakeCase("URLValue"));
```
test case for a string that has a leading underscore?

stephentoub replied:
(and multiple leading underscores)

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,141 @@
+
+        [Theory]
+        [MemberData(nameof(NamesToTest))]
+        public static void Convert_DictionaryKey_MatchesNewtonsoft(string name)
```
For  completeness, can we also add tests where the `JsonSerializer` is used to serialize POCOs with this naming policy passed in as `JsonSerializerOptions`?

Include some round-tripping (deserialize/serialize/deserialize), similar to https://github.com/dotnet/corefx/pull/40003/files#diff-b6e2569f6a455af110a90956eaa53eb8R29.

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,141 @@
+        [MemberData(nameof(NamesToTest))]
+        public static void Convert_DictionaryKey_MatchesNewtonsoft(string name)
+        {
+            JsonNamingPolicy policy = JsonNamingPolicy.SnakeCase;
```
Also, I believe (as far as I can recall), the SnakeCase naming policy will not work with round-tripping enum values as strings. Can you try that out and see what the current behavior is of this policy when combined with StringEnumConverter? CamelCase happened to work because we used ignorecase.

https://github.com/dotnet/corefx/blob/1338e4e8a542c3b3e20d99e544e09860f23e99ea/src/System.Text.Json/src/System/Text/Json/Serialization/JsonStringEnumConverter.cs#L42-L48

https://github.com/dotnet/corefx/blob/1338e4e8a542c3b3e20d99e544e09860f23e99ea/src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs#L59-L67

We may need to update that behavior if it isn't supported, or document it.
cc @steveharter, @JeremyKuhne

stephentoub commented:
```
    System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Convert_SpecifiedName_MatchesExpected(expected: "__already_snake_case", name: "already_snake_case") [FAIL]
      Assert.Equal() Failure
                 (pos 0)
      Expected: __already_snake_case
      Actual:   already_snake_case
                 (pos 0)
      Stack Trace:
        /_/src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs(101,0): at System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Convert_SpecifiedName_MatchesExpected(String expected, String name)
    System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Convert_SpecifiedName_MatchesExpected(expected: "_already_snake_case", name: "already_snake_case") [FAIL]
      Assert.Equal() Failure
                 (pos 0)
      Expected: _already_snake_case
      Actual:   already_snake_case
                 (pos 0)
      Stack Trace:
        /_/src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs(101,0): at System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Convert_SpecifiedName_MatchesExpected(String expected, String name)
    System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Serialize_RoundTipping_MatchesOriginal [FAIL]
      Assert.Equal() Failure
                              (pos 13)
      Expected: {"my_int16":1}
      Actual:   {"my_int16":1,"my_int32":0,"my_int64":0,"my_u_int16":0úúú
                              (pos 13)
      Stack Trace:
        /_/src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs(111,0): at System.Text.Json.Serialization.Tests.SnakeCaseUnitTests.Serialize_RoundTipping_MatchesOriginal()
```

ahsonkhan commented:
We should use this new naming policy within the tests being added as part of https://github.com/dotnet/corefx/pull/41691 (once that's merged).

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,125 @@
+        //
+        [InlineData("hi_this_is_text_time_to_test", "Hi!! This is text. Time to test.")]
+        //
+        //
```
Remove extra line

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,125 @@
+        [Fact]
+        public static void Serialize_RoundTipping_MatchesOriginal()
+        {
+            var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.SnakeCase };
```
Add tests where we use `JsonNamingPolicy.SnakeCase` to set `DictionaryKeyPolicy` as well.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,70 @@
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
+                        if (previousCategory == UnicodeCategory.SpaceSeparator)
+                            builder.Append('_');
```
nit: Add braces around the if since this is in the nested scope.

From https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md#c-coding-style
> We use [Allman style](http://en.wikipedia.org/wiki/Indent_style#Allman_style) braces, where each brace begins on a new line. A single line statement block can go without braces but the block must be properly indented on its own line and must not be nested in other statement blocks that use braces (See rule 17 for more details). One exception is that a `using` statement is permitted to be nested within another `using` statement by starting on the following line at the same indentation level, even if the nested `using` contains a controlled block.

YohDeadfall replied:
Ah, I see.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,70 @@
+
+                    default:
+                        if (previousCategory != null)
+                            previousCategory = UnicodeCategory.SpaceSeparator;
```
Same here (braces).

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,70 @@
+                {
+                    case UnicodeCategory.UppercaseLetter:
+                    case UnicodeCategory.TitlecaseLetter:
+                        if (previousCategory == UnicodeCategory.SpaceSeparator ||
```
nit: Put brackets around the && conditions to make the ordering of operations clear.

YohDeadfall commented:
@stephentoub Yep, saw it, but had no time yet to care about this.

YohDeadfall commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,70 @@
+                    continue;
+                }
+
+                UnicodeCategory currentCategory = char.GetUnicodeCategory(currentChar);
```
@ahsonkhan Out of curiosity, when method already contains the return type name is it okay to use `var` on the left side?

ahsonkhan replied:
I don't think so (but it is a bit of a grey zone). Use `var` when the RHS has `new` or otherwise it is clear what the type is from reading that line (without having prior knowledge of what the calling API is doing - such as an explicit cast).

> We only use `var` when it's obvious what the variable type is

I can see the argument that GetX methods return X and hence it is clear, so this comes down to what is considered "obvious" (and probably personal preference). What if GetX was returning an interface/abstract type (like IX)?

ahsonkhan commented:
@YohDeadfall - would you be able to address the testing feedback and get this PR ready to merge today? If not, are you planning on restarting it once the dotnet/runtime repo is open?

Let me know, especially if you don't have the time atm, and would like one of us to carry it forward. Otherwise, we can certainly wait a bit :)

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,74 @@
+                        break;
+
+                    case UnicodeCategory.LowercaseLetter:
+                    case UnicodeCategory.DecimalDigitNumber:
```
What's the rational for special casing `DecimalDigitNumber`? Can you share/add a test case?

ahsonkhan replied:
This category contains a bunch of characters. Is 0-9 sufficient?
https://www.fileformat.info/info/unicode/category/Nd/list.htm

YohDeadfall replied:
Tests already exists for numbers in names. One of them you have used in benchmarks (:

Since the policy should put underscores and not remove any characters and numbers including non-ASCII, this category should be used.

ahsonkhan replied:
Ah yes, I see the test case in the bug https://github.com/npgsql/npgsql/issues/2152.

`ABC123 -> abc123` clearly indicates the need for using 0-9 (`ab_c123` is clearly wrong). I see the numbers tests now :)

To be explicit, having non-ASCII digit tests would be good too (i.e. pick others from the Nd list).

YohDeadfall replied:
What's about non-decimal numbers, math symbols and other characters? Maybe invert the logic and write everything except punctuation and spaces?

/cc @GrabYourPitchforks

ahsonkhan replied:
> everything except punctuation and spaces

That makes sense to me.

Not sure how accurate this is, but it is a data point (from https://capitalizemytitle.com/camel-case/):
> Basic Snake Case Capitalization Rules
All letters are lowercase.
All spaces between words are filled with underscores.
Remove all punctuation.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,74 @@
+                switch (currentCategory)
+                {
+                    case UnicodeCategory.UppercaseLetter:
+                    case UnicodeCategory.TitlecaseLetter:
```
I don't see why we need to special case TitlecaseLetter. Can we ignore them?

https://www.fileformat.info/info/unicode/category/Lt/list.htm

YohDeadfall replied:
Do you mean that they cannot be lowercased? I haven't worked with Unicode a lot, so it's out of my knowledge.

ahsonkhan replied:
I am also not super familiar with all the categories. I am asking if you had a test case or scenario that motivated including `TitlecaseLetter` in the implementation to begin with. It looks like they do have lower-case counterparts. Again, adding tests where we exercise this `UnicodeCategory` more would be useful. I believe, if we remove this switch case, all the existing tests would still pass, which means there is a test gap.

@GrabYourPitchforks - do you have some thoughts on the unicode categories we are special casing here for snake casing?

ahsonkhan replied:
I am primarily asking for adding more tests so that we can more confidently optimize the implementation later. I want to avoid removing some code path in the future to improve perf that would end up breaking functionality.

YohDeadfall replied:
Checked. They all can be lowered.

ahsonkhan commented:
@YohDeadfall - can you please retain the update commits so that it is easier to see a diff and review incremental changes? We squash the commits into one when merging anyway, but having multiple commits makes it easier to review just the recent changes (rather than having to review the whole PR again).

YohDeadfall commented:
> would you be able to address the testing feedback and get this PR ready to merge today? If not, are you planning on restarting it once the dotnet/runtime repo is open?

Addressed review. If there are some things which should be fixed, will do it in PR to the new repo.

> can you please retain the update commits so that it is easier to see a diff and review incremental changes?

Not a problem, will do. Just get used to it in the Npgsql repo and others (:

ahsonkhan commented:
> "II "

Revisiting the input strings with trailing space characters, maybe we shouldn't remove those, but rather substitute with trailing `_` instead.

E.g.: https://capitalizemytitle.com/camel-case/
"ABC123 " -> "abc123_"

https://html-cleaner.com/case-converter/

Regarding numbers and other special test cases, maybe we can look at what pylint does in python:
https://docs.pylint.org/en/1.6.0/options.html

YohDeadfall commented:
> Revisiting the input strings with trailing space characters, maybe we shouldn't remove those, but rather substitute with trailing _ instead. Thoughts?

The current behavior (in Npgsql) was kept for compatibility with the previous implementations which was taken from JSON.NET, so the policy ignores trailing white spaces too. Don't think that it's correct because this:
* increases collision probability,
* inconsistent with itself since it doesn't ignore non-tailing spaces.

Another thing I'm considered on is ignoring some character groups, for example punctuation. Camel case translator prevents all characters in an input string, while this is not, so... Would it be better to change the logic to the following:
* replace all non-digit and non-alphabet characters by underscores,
* put underscores between words if there is no spaces between.

The resulting sting will have the same length as the input. Therefore, there is a less chance to meet a collision.

Recently we got an issue about bad naming in some cases (see npgsql/npgsql#2734). but as I see all translators use the same logic as the policy now, i.e. doesn't add an underscore between digits and characters.

YohDeadfall committed:
Removed already covered test case

YohDeadfall committed:
Improved tests for non-ASCII characters

YohDeadfall committed:
Added support for surogates

ahsonkhan commented:
> put underscores between words if there is no spaces between.

You mean if there **are** spaces between them, right?

> The resulting sting will have the same length as the input. Therefore, there is a less chance to meet a collision.

That seems like a good principle to aim towards.

ahsonkhan commented:
Some more references to find precedence of behavior.

JavaScript:
https://www.npmjs.com/package/change-case
- Ignores leading/trailing whitespace
- Collapse multiple underscores into one
- Each capital letter is considered one "word"
- Numbers are only treated as separate words if followed by capital letter
```js
var changeCase = require('change-case')
console.log( changeCase.snakeCase('foo_bar_baz')) // => "foo_bar_baz"
console.log( changeCase.snakeCase('One Word another Word')) // => "one_word_another_word"
console.log( changeCase.snakeCase('One Word another Word')) // => "one_word_another_word" <= special white space https://www.fileformat.info/info/unicode/char/2008/index.htm
console.log( changeCase.snakeCase('One 1Word 1another1 1Word')) // => "one_1_word_1another1_1_word"
console.log( changeCase.snakeCase('One ٤Word ٤another٤ ٤Word')) // => "one_٤_word_٤another٤_٤_word"
console.log( changeCase.snakeCase('FooBarBaz')) // => "foo_bar_baz"
console.log( changeCase.snakeCase('  FooBarBaz  ')) // => "foo_bar_baz" <= special white space https://www.fileformat.info/info/unicode/char/2003/index.htm
console.log( changeCase.snakeCase('foo__bar_baz')) // => "foo_bar_baz"
console.log( changeCase.snakeCase(' ii ')) // => "ii"
console.log( changeCase.snakeCase('   ii123 ')) // => "ii123"
console.log( changeCase.snakeCase('XMLHtppRequest')) // => "xml_htpp_request"
console.log( changeCase.snakeCase('ABC123AnotherWord')) // => "abc123_another_word"
console.log( changeCase.snakeCase('ǅǅǅǈᾊabcῌ')) // => "ǆǆǆǉᾂabcῃ"
console.log( changeCase.snakeCase('ABC٥٤߅')) // => "abc٥٤߅"
console.log( changeCase.snakeCase('ABC٥٤߅A')) // => "abc٥٤߅_a"
```

Python:
https://pypi.org/project/stringcase/
- Adds leading/trailing underscores for whitespace
- Doesn't collapse multiple underscores into one
- Each capital letter is separated by underscore
- Numbers are only treated as separate words if followed by capital letter
- Seems a bit buggy for non-ascii (doesn't lower-case all letters)
- Another buggy behavior: Adds two underscores for whitespace before a capital letter
```py
import stringcase
print(stringcase.snakecase('foo_bar_baz')) # => "foo_bar_baz"
print(stringcase.snakecase('One Word another Word')) # => "one__word_another__word"
print(stringcase.snakecase('One Word another Word')) # => "one__word_another__word" <= special white space https://www.fileformat.info/info/unicode/char/2008/index.htm
print(stringcase.snakecase('One 1Word 1another1 1Word')) # => "one_1_word_1another1_1_word"
print(stringcase.snakecase('One ٤Word ٤another٤ ٤Word')) # => "one_٤_word_٤another٤_٤_word"
print(stringcase.snakecase('FooBarBaz')) # => "foo_bar_baz"
print(stringcase.snakecase('  FooBarBaz  ')) # => "___foo_bar_baz__" <= special white space https://www.fileformat.info/info/unicode/char/2003/index.htm
print(stringcase.snakecase('foo__bar_baz')) # => "foo__bar_baz"
print(stringcase.snakecase(' ii ')) # => "_ii_"
print(stringcase.snakecase('   ii123 ')) # => "___ii123_"
print(stringcase.snakecase('XMLHtppRequest')) # => "x_m_l_htpp_request"
print(stringcase.snakecase('ABC123AnotherWord')) # => "a_b_c123_another_word"
print(stringcase.snakecase('ǅǅǅǈᾊabcῌ')) # => "ǆǅǅǈᾊabcῌ"
print(stringcase.snakecase('ABC٥٤߅')) # => "a_b_c٥٤߅"
print(stringcase.snakecase('ABC٥٤߅A')) # => "a_b_c٥٤߅_a"
```

ahsonkhan commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,157 @@
+        [InlineData("u_ppe_r_case", "UPpeR Case")]
+        [InlineData("u_ppe_r_c_a_se", "UPpeR cASe")]
+        //
+        [InlineData("ä", "ä")]
```
Non-ascii tests like these should be written using escaped/hex characters.

> When including non-ASCII characters in the source code use Unicode escape sequences (\uXXXX) instead of literal characters. Literal non-ASCII characters occasionally get garbled by a tool or editor.

See: 
https://github.com/dotnet/corefx/blob/5cee7c97d602f294e27c582d4dab81ec388f1d7b/src/System.Text.Json/tests/Utf8JsonReaderTests.cs#L1993

So, for this case (same for the special "42" below):
```suggestion
        [InlineData("\u00E4", "\u00E4")]
```

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,77 @@
+                        }
+                        break;
+
+                    case UnicodeCategory.Surrogate:
```
Consider adding tests for invalid surrogate characters too (which is possible by doing a substring in between a surrogate pair of characters or crafting one using chars).
So: 
- a string containing a high-surrogate without a low-surrogate following it
- a string containing a low-surrogate first before a high-surrogate

YohDeadfall replied:
Already have one of them. It's 42.

ahsonkhan replied:
The special `42` contains **valid** surrogate pairs (https://www.fileformat.info/info/unicode/char/1d7dc/index.htm).

UTF-16 (hex) | 0xD835 0xDFDC (d835dfdc)
-- | --

I am talking about a string that is **invalid** (i.e. doesn't contain the correct pair of surrogate characters).

For example:
https://github.com/dotnet/corefx/blob/5cee7c97d602f294e27c582d4dab81ec388f1d7b/src/System.Text.Json/tests/JsonDocumentTests.cs#L3545-L3546

YohDeadfall replied:
Missed word "invalid", sorry. Will do (:

YohDeadfall committed:
Added tests for invalid values

YohDeadfall commented:
Agreed with you, but should we follow the rule that adds an underscore after numbers? If so `Xml2Json` will be translated as `xml2_json` instead of `xml2json`.

khellang commented:
> should we follow the rule that adds an underscore after numbers?

I'd say either `xml_2_json` or `xml2json`, but not `xml2_json`.

YohDeadfall commented:
@ygoe shown an example which doesn't look well without underscores (`En13757Index` is translated as `en13757index`), so prefixing numbers sounds good to me.

khellang commented:
I think @ahsonkhan's observations from the two linked implementations in Python and Node makes sense:

> Numbers are only treated as separate words if followed by capital letter

YohDeadfall commented:
The `change-case`  package splits the passed string into words and then adds a separator between words. Therefore, it collapses multiple underscores into one and ignores leading/trailing spaces.

Here is a map of characters which should appear in a resulting string using the package:

| Unicode category                | Should be in resulting string |
|---------------------------------|-------------------------------|
| Cc (other, control)             | No                            |
| Cf (other, format)              | No                            |
| Cn (other, not assigned)        | No                            |
| Co (other, private use)         | No                            |
| Cs (other, surrogate)           | No                            |
| Ll (letter, lowercase)          | Yes, see exceptions           |
| Lm (letter, modifier)           | Yes, all                      |
| Lo (letter, other)              | Yes, see exceptions           |
| Lt (letter, titlecase)          | Yes, all                      |
| Lu (letter, uppercase)          | Yes, see exceptions           |
| Mc (mark, spacing combining)    | Yes, see exceptions           |
| Me (mark, enclosing)            | No                            |
| Mn (mark, nonspacing)           | No                            |
| Nd (number, decimal digit)      | Yes, all                      |
| Nl (number, letter)             | Yes, all                      |
| No (number, other)              | Yes, see exceptions           |
| Pc (punctuation, connector)     | No                            |
| Pd (punctuation, dash)          | No                            |
| Pe (punctuation, close)         | No                            |
| Pf (punctuation, final quote)   | No                            |
| Pi (punctuation, initial quote) | No                            |
| Po (punctuation, other)         | No                            |
| Ps (punctuation, open)          | No                            |
| Sc (symbol, currency)           | No                            |
| Sk (symbol, modifier)           | No                            |
| Sm (symbol, math)               | No                            |
| So (symbol, other)              | No                            |
| Zl (separator, line)            | No                            |
| Zp (separator, paragraph)       | No                            |
| Zs (separator, space)           | No                            |

### Included into an output

| Unicode category             | Characters |
|------------------------------|------------|
| Mc (mark, spacing combining) | `U+1885`, `U+1886` |

### Excluded from an output

| Unicode category       | Characters |
|------------------------|------------|
| Ll (letter, lowercase) | `U+0560`, `U+0588`, `U+1C80` - `U+A7AF`, `U+A7B9` |
| Lo (letter, other)     | `U+05EF`, `U+0860`, `U+0861` - `U+08B6`, `U+08B7` - `U+09FC`, `U+0C80`, `U+0D54`, `U+0D55` - `U+1878`, `U+312E` - `U+9FD6`, `U+9FD7` - `U+A8FE` |
| Lu (letter, uppercase) | `U+1C90` - `U+1CBD`, `U+1CBE` - `U+A7AE`, `U+A7B8` |
| No (number, other)     | `U+0D58` - `U+0D5E`, `U+0D76` - `U+0D78` |

## Questions

### Should surrogates be included?

_Pros:_ Compatibility with the existing implementations. Lowers collision probability
_Cons:_ Undetermined.

### Should listed above characters be excluded or included?

Some of them ansuvara (randomly checked) so there is no reason to exclude this symbols since they behave like variations, but phonetic. To make the final decision each character should be checked.

### Should all excluded letters be replaced by an underscore?

_Pros:_ Lowers collision probability.
_Cons:_ The result would look uglier in the case of many replacing characters.

### Should leading/trailing replacing characters be ignored?

_Pros:_ Compatibility with the existing implementations.
_Cons:_ Lowers collision probability.

### Should multiple underscores be collapsed?

_Pros:_ Compatibility with the existing implementations only when letters get replaced. Underscores in the original string should be preserved.
_Cons:_ Lowers collision probability.

### Should an underscore be added after a number?

_Pros:_ Compatibility with the existing implementations.
_Cons:_ Some edge cases like `xml2json` will be properly handled.

YohDeadfall commented:
@ahsonkhan To make things simple I researched the npm package and added the table you can see above. Let's analyze it and answer the questions first so we won't spent time on adding test and changing the behavior multiple times. After that I will write a final and optimized version of the policy.

khellang commented:
Question: How does all of this compare to the existing behavior of camelCase naming?

It would certainly be interesting to use the same word-splitting behavior for all naming conventions.
That would make it really easy to add new conventions, as it would only be a matter of changing casing and adding separators between words:

| Name | First | Rest | Separator |
|------|-------|------|-----------|
| PascalCase | Capitalize | Capitalize | None |
| camelCase | LowerCase | Capitalize | None |
| snake_case | LowerCase | LowerCase | `_` |
| SCREAMING_SNAKE_CASE | UpperCase | UpperCase | `_` |
| camel_Snake_Case | LowerCase | Capitalize | `_` |
| Pascal_Snake_Case | Capitalize | Capitalize | `_` |
| kebab-case | LowerCase | LowerCase | `-` |
| SCREAMING-KEBAB-CASE | UpperCase | UpperCase | `-` |

YohDeadfall commented:
This is how the npm works, but the camel case policy (from the new JSON serializer) just lowers the first characters and does nothing more due to performance reasons. So an user controls all characters except the first one if it's a letter. Otherwise, the user controls all of them.

The snake case policy isn't limited to the first, so it's a problem.

khellang commented:
> This is how the npm works

This is what [camel-snake-kebab](https://github.com/clj-commons/camel-snake-kebab/blob/0e268da03ccd48a3c690230c76d905994908c71a/src/camel_snake_kebab/core.cljc#L49-L55) in Clojure does as well. It's very consistent.

> the camel case policy (from the new JSON serializer) just lowers the first characters and does nothing more due to performance reasons.

How is this performance critical? Is the result of the transformation not cached?

Anyway, I guess the ship has sailed for camelCase as it would be a major breaking change to switch behavior now, but for this and (the inevitable ask for) future conventions, it would be nice if they used the same strategy. I guess this could be refactored out if it becomes relevant later.

stephentoub commented:
>  Is the result of the transformation not cached?

https://github.com/dotnet/corefx/pull/41354#discussion_r344367829

YohDeadfall commented:
The serializer caches only property and class names, but not dictionary keys. Having a pool would help only if it supports LRU strategy.

YohDeadfall commented:
> Anyway, I guess the ship has sailed for camelCase as it would be a major breaking change to switch behavior now

This is major problem, but as I know the serializer isn't very popular due to limitations, a lot of them. So if a breakage would be introduced, it won't affect to many people. Taking into account that any policy is mostly used for members and classes, the change won't hurt so much, but will give a positive impact.

maryamariyan commented:
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

layomia commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,160 @@
+                actual.SomeIntProperty);
+        }
+
+        private class NamingPolictyTestClass
```
typo `NamingPolictyTestClass` -> `NamingPolicyTestClass`

layomia commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,160 @@
+            var expected = new NamingPolictyTestClass { SomeIntProperty = 42 };
+            var actual = JsonSerializer.Deserialize<NamingPolictyTestClass>(JsonSerializer.Serialize(expected, options), options);
+
+            Assert.Equal(
+                expected.SomeIntProperty,
+                actual.SomeIntProperty);
```
This is short enough to be one line, no?

layomia commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,160 @@
+        [Fact]
+        public static void SerializeDictionary_RoundTipping_MatchesOriginal()
+        {
+            var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.SnakeCase };
```
This should be `DictionaryKeyPolicy = JsonNamingPolicy.SnakeCase`

layomia commented on src/System.Text.Json/tests/Serialization/SnakeCaseUnitTests.cs:
```diff
@@ -0,0 +1,160 @@
+            var options = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.SnakeCase };
+
+            string expected = @"{""some_int_property"":42}";
+            string actual = JsonSerializer.Serialize(JsonSerializer.Deserialize<Dictionary<string, int>>(expected, options), options);
```
`DictionaryKeyPolicy` is only applied on serialization. The way to test this is to have a dictionary with keys that are some other casing like Pascal case, and make sure the serialized content is snake case.
